apps/ligo-virgo/src/pages/workspace/docs/components/toc/TOC.tsx

```diff
@@ -0,0 +1,176 @@
+                padding: '6px 12px',
+                fontWeight: '600',
+                fontSize: '16px',
+                borderTop: '0.5px solid #E0E6EB',
+                borderBottom: '0.5px solid #E0E6EB',
+                color: isActive ? '#3E6FDB' : '#98ACBD',
+            };
+        }
+
+        return {};
+    }
+);
+
+const StyledItem = styled('div')(props => {
+    return {
+        overflow: 'hidden',
+        textOverflow: 'ellipsis',
+        whiteSpace: 'nowrap',
+    };
+});
+
+const TOCContext = createContext(null);
+
+interface Props {
+    children: ReactNode;
+    editor: BlockEditor;
+}
+
+const TOCItem = props => {
+    const { activeBlockId, onClick } = useContext(TOCContext);
+    const { id, type, text } = props;
+    const isActive = id === activeBlockId;
+
+    return (
+        <StyledTOCItem
+            key={id}
+            isActive={isActive}
+            type={type}
+            onClick={() => onClick(id)}
+        >
+            <StyledItem>{text}</StyledItem>
+        </StyledTOCItem>
+    );
+};
+
+const renderTOCContent = tocDataSource => {
+    return (
+        <>
+            {tocDataSource.map(tocItem => {
+                if (tocItem?.length) {
+                    return renderTOCContent(tocItem);
+                }
+
+                const { id, type, text } = tocItem;
+
+                return <TOCItem key={id} id={id} type={type} text={text} />;
+            })}
+        </>
+    );
+};
+
+export const TOC = (props: Props) => {
+    const { editor } = props;
+    const { page_id } = useParams();
+    const [tocDataSource, setTocDataSource] = useState<TocType[]>([]);
+    const [activeBlockId, setActiveBlockId] = useState('');
+
+    const updateTocDataSource = useCallback(async () => {
+        /* page listener: trigger update-notice when add new group */
+        const pageAsyncBlock = (await editor.getBlockByIds([page_id]))?.[0];
+        if (!listenerMap.has(pageAsyncBlock.id)) {
+            listenerMap.set(
+                pageAsyncBlock.id,
+                pageAsyncBlock.onUpdate(updateTocDataSource)
+            );
+        }
+
+        /* block listener: trigger update-notice when change block content */
+        const { children = [] } =
+            (await editor.queryByPageId(page_id))?.[0] || {};
+        const asyncBlocks = (await editor.getBlockByIds(children)) || [];
+        const { tocContents } = await getContentByAsyncBlocks(
+            asyncBlocks,
+            updateTocDataSource
+        );
+
+        /* toc: flat content */
+        const tocDataSource = getPageTOC(asyncBlocks, tocContents);
+        setTocDataSource(tocDataSource);
+
+        /* remove listener when unmount component */
+        return destroyEventList;
+    }, [editor, page_id]);
+
+    /* init toc and add page/block update-listener & unmount-listener */
+    useEffect(() => {
+        (async () => {
+            await updateTocDataSource();
+        })();
+    }, [updateTocDataSource]);
```

Comment:
Why not use useEffect directly?

```tsx
useEffect(() => {
  const updateTocDataSource = async () => {
    // ...
  }
  // call fn immediately
  updateTocDataSource()
}, [editor, page_id])

```

Reply:
Contract them together is Ok, but I think layered:

In TOC,  `useEffect` will trigger only once, but `updateTocDataSource` will be exec when:
- page-content changed(example: add new Group);
- block-content changed(example: input something in a block);

when something is error, can simple exclude effect about Effect Hook, not need to think whether something dependencies had change.